lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -679,6 +688,220 @@ void ComponentOp::getAsmBlockArgumentNames(
+  p << " -> ";
+  printPortDefList(getOutputPortInfo());
+
+  p << " ";
+  p.printRegion(getBody(), /*printEntryBlockArgs=*/false,
+                /*printBlockTerminators=*/false,
+                /*printEmptyBlock=*/false);
+
+  SmallVector<StringRef> elidedAttrs = {"portAttributes", "portNames",
+                                        "portDirections", "sym_name",
+                                        ComponentOp::getTypeAttrName()};
+  p.printOptionalAttrDict((*this)->getAttrs(), elidedAttrs);
+}
+
+ParseResult CombComponentOp::parse(OpAsmParser &parser,
+                                   OperationState &result) {
+  using namespace mlir::function_interface_impl;
+
+  StringAttr componentName;
+  if (parser.parseSymbolName(componentName,
+                             ::mlir::SymbolTable::getSymbolAttrName(),
+                             result.attributes))
+    return failure();
+
+  SmallVector<mlir::OpAsmParser::Argument> ports;
+
+  SmallVector<Type> portTypes;
+  if (parseComponentSignature(parser, result, ports, portTypes))
+    return failure();
+
+  // Build the component's type for FunctionLike trait. All ports are listed
+  // as arguments so they may be accessed within the component.
+  auto type = parser.getBuilder().getFunctionType(portTypes, /*results=*/{});
+  result.addAttribute(CombComponentOp::getTypeAttrName(), TypeAttr::get(type));
+
+  auto *body = result.addRegion();
+  if (parser.parseRegion(*body, ports))
+    return failure();
+
+  if (body->empty())
+    body->push_back(new Block());
+
+  if (parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  return success();
+}
+
+LogicalResult CombComponentOp::verify() {
+  // Verify there is exactly one of each the wires and control operations.
+  auto wIt = getBodyBlock()->getOps<WiresOp>();
+  if (std::distance(wIt.begin(), wIt.end()) != 1)
+    return emitOpError("requires exactly one 'calyx.wires' op.");
+
+  // Verify the component actually does something: has continuous assignments.
+  bool hasNoAssignments =
+      getWiresOp().getBodyBlock()->getOps<AssignOp>().empty();
+  if (hasNoAssignments)
+    return emitOpError(
+        "The component currently does nothing. It needs to either have "
+        "continuous assignments in the Wires region or control constructs in "
+        "the Control region.");
+
+  // Check that all cells are combinational
+  auto cells = getOps<CellInterface>();
+  for (auto cell : cells) {
+    if (!cell.isCombinational())
+      return emitOpError()
+             << "Combinational component contains non-combinational cell "
+             << cell.instanceName();
+  }
+
+  // Check that the component has no groups
+  auto groups = getWiresOp().getOps<GroupOp>();
+  auto combGroups = getWiresOp().getOps<CombGroupOp>();
+
+  if (!groups.empty())
+    return emitOpError() << "Combinational component contains group "
+                         << (*groups.begin()).getSymName();
+
+  if (!combGroups.empty())
```

Comment:
This might be initially confusing for someone who is not familiar with Calyx combinational groups. I'd suggest adding a comment here with a link to the Calyx documentation on why combinational groups aren't allowed in combinational components:

https://docs.calyxir.org/lang/ref.html#comb-group-definitions